sgx/epid-types/src/update_info.rs

```diff
@@ -0,0 +1,224 @@
+
+const UCODE_START: usize = 0;
+const UCODE_END: usize = UCODE_START + INTEL_U32_SIZE;
+const CSME_START: usize = UCODE_END;
+const CSME_END: usize = CSME_START + INTEL_U32_SIZE;
+const PSW_START: usize = CSME_END;
+const PSW_END: usize = PSW_START + INTEL_U32_SIZE;
+
+/// The size of the [UpdateInfo]'s x64 representation, in bytes.
+pub const UPDATE_INFO_SIZE: usize = PSW_END;
+
+/// An update info structure, describing which parts (if any) of the platform's
+/// [Trusted Computing Base](https://en.wikipedia.org/wiki/Trusted_computing_base)
+/// require updates to remain secure.
+#[derive(Default)]
+#[repr(transparent)]
+pub struct UpdateInfo(sgx_update_info_bit_t);
+
+impl_ffi_wrapper_base! {
+    UpdateInfo, sgx_update_info_bit_t, UPDATE_INFO_SIZE;
+}
+
+impl_serialize_to_x64! {
+    UpdateInfo, UPDATE_INFO_SIZE;
+}
+
+impl Debug for UpdateInfo {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let ucode_update = self.0.ucodeUpdate;
+        let csme_fw_update = self.0.csmeFwUpdate;
+        let psw_update = self.0.pswUpdate;
+        write!(
+            f,
+            "UpdateInfo {{ ucodeUpdate: i32({}), csmeFwUpdate: i32({}), pswUpdate: i32({}) }}",
+            ucode_update, csme_fw_update, psw_update,
+        )
+    }
+}
+
+impl Display for UpdateInfo {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        let ucode_update = self.0.ucodeUpdate;
+        let csme_fw_update = self.0.csmeFwUpdate;
+        let psw_update = self.0.pswUpdate;
+        write!(
+            f,
+            "Microcode {}, Management Engine Firmware {}, Platform Services {}",
+            ucode_update, csme_fw_update, psw_update
+        )
+    }
+}
+
+impl FfiWrapper<sgx_update_info_bit_t> for UpdateInfo {}
+
+impl From<&sgx_update_info_bit_t> for UpdateInfo {
+    fn from(src: &sgx_update_info_bit_t) -> UpdateInfo {
+        Self(sgx_update_info_bit_t {
+            ucodeUpdate: src.ucodeUpdate,
+            csmeFwUpdate: src.csmeFwUpdate,
+            pswUpdate: src.pswUpdate,
+        })
+    }
+}
+
+impl FromX64 for UpdateInfo {
+    type Error = EncodingError;
+
+    fn from_x64(src: &[u8]) -> Result<UpdateInfo, EncodingError> {
+        Ok(Self(sgx_update_info_bit_t {
+            ucodeUpdate: i32::from_le_bytes(src[0..4].try_into()?),
+            csmeFwUpdate: i32::from_le_bytes(src[4..8].try_into()?),
+            pswUpdate: i32::from_le_bytes(src[8..12].try_into()?),
+        }))
+    }
+}
+
+impl Hash for UpdateInfo {
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        "mc_sgx_epid_types::update_info::UpdateInfo".hash(hasher);
+        let mut value = self.0.ucodeUpdate;
+        value.hash(hasher);
+        value = self.0.csmeFwUpdate;
+        value.hash(hasher);
+        value = self.0.pswUpdate;
+        value.hash(hasher);
+    }
+}
+
+impl Ord for UpdateInfo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let left_ucode = self.0.ucodeUpdate;
```

Comment:
I think this can be simplified into `(self.0.ucodeUpdate, self.0.csmeFwUpdate, self.0.pswUpdate).cmp(&(self.other.ucodeUpdate, self.other.csmeFwUpdate, other.0.pswUpdate))`

Reply:
Oooh yeah, I'll check that out.